### PR TITLE
pja-es-filter-item

### LINKED ISSUE #11: 9. As a user, I want to filter my shopping list to make it easier to locate an item in the list
AC:

* Display a text field at the top of the shopping list
* As the user types into the field, the list narrows to include only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field

### PR DESCRIPTION
**Pull request by espain16**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

This code adds a search functionality to the app, which enables the user to search through their shopping list.

The code adds a filter component that renders on the shopping list component. When the user types an item into the search bar an event handler is triggered. The event takes that item and double checks if it's in the list using `.filter`, and `.toLowerCase` accounts for any difference in the user input. The user also has the ability to clear  the search bar if a new search is needed. 

## Related Issue
closes #11 

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

PR 5:10:2020

### After

PR 5:10:2020
PR 5:10:2020 2

## Testing Steps / QA Criteria

1) Navigate to the website.  You should see an option to "Join a list."  This is where the user would enter a token to join a shopping list.  For this test, the user token is "annex hagen veldt." 
2) After entering the user token and clicking SUBMIT, you will be taken to the "Shopping List" screen.
3) Use the SEARCH bar to check the functionality of the filter.

### CODE REVIEW

File: src/components/Filter.js [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+    console.log('Hey you typed something');
+  }
 
-  // use onChange ini a function
-  function handleTextChange(event) {}
-
-  return <input type="text" placeholder="Search"></input>;
+  return (
+    <input onChange={handleTextChange} type="text" placeholder="Search"></input>
+  );
```
Comment (laurenmbeatty) [inline src/components/Filter.js:14]: These incremental commits are awesome.  💯

File: src/components/shopping.js [hunk 2]
```diff
@@ -1,13 +1,35 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { Link } from 'react-router-dom';
 import ListItem from './ListItem';
+import Filter from './Filter';
 
 function Shopping(props) {
+  const [filterString, setFilterString] = useState('');
+  const [filteredList, setFilteredList] = useState(props.list);
+  useEffect(() => {
```
Comment (skillitzimberg) [inline src/components/shopping.js:9]: Cool! You got to use `useEffect`!

File: src/components/shopping.js [hunk 3]
```diff
@@ -1,13 +1,35 @@
...
+  const handleTextChange = event => {
+    setFilterString(event.target.value);
+    const newList = props.list.filter(item =>
+      item.name.toLowerCase().includes(event.target.value.toLowerCase()),
+    );
+    setFilteredList(newList);
+  };
+  const handleClear = () => {
+    setFilterString('');
+    props.list.length > 0 && setFilteredList(props.list);
```
Comment (skillitzimberg) [inline src/components/shopping.js:21]: This is interesting! I'm not sure what it does . . . I'm going to have to learn this one.

**Review by skillitzimberg — APPROVED** [review]
Looks good.

**Review by SaraSweetie — APPROVED** [review]
This works very well! You gals did a good job.

**Review by laurenmbeatty — APPROVED** [review]
👍

**Pull request merged**